### PR TITLE
Add Comment Section To Repository

### PR DESCRIPTION
**Pull request by WittyWhiscash**
This PR aims to add a robust comment section to the repository, along with moderation tools to monitor and delete/restore comments.

image

In order to achieve this, a new table was added to the database which tracks comments and ratings on a map.

Once that was in place, a new class for comments was added and integrated into the website. Adjustments to the map class were made to automatically load in comments up to a certain limit based on the number passed (or not passed) into the methods.

Once that was in place, a new set of admin tools was integrated into the website to monitor and remove problematic comments.

image

In order for this to run properly, you'll have to migrate the database and add the new mapcomments table. In addition, I would also define new setting into the config.php named AMNT_FLAGS_BEFORE_REMOVE, with an integer as the value. This will determine the amount of reports before a comment is brought for moderator review.

### CODE REVIEW

File: admin.php
Comment (14ercooper) [inline admin.php]: I think I'd prefer another field on the comment for indicating "Admin Approved" instead of simply resetting the flag count, otherwise people could spam report a given person's comments endlessly and constantly remove them, even if an admin has already reviewed and approved the comment.

File: templates/submitCommentForm.php
Comment (14ercooper) [inline templates/submitCommentForm.php]: Some bot is going to try to paste an entire novel into this field if there's no character limit, comments should probably have a limit that's quite a bit smaller than the limits on a map post.

File: index.php
Comment (14ercooper) [inline index.php]: I feel like I'd want the increment to be done as an atomic operation at the database level, otherwise if 3 people report a spam comment at the same time, it could be possible for the flag count to only get incremented once due to a race condition.

**Review by 14ercooper — CHANGES_REQUESTED** [review]
Just a few small things I noticed I'd like changed, mostly around handling edge cases.
There's also a few merge conflicts on the PR due to some bugfixes to older I pushed up in the interim that I'll need resolved before I merge in this PR.

File: templates/submitCommentForm.php
Comment (WittyWhiscash) [reply templates/submitCommentForm.php]: I've adjusted the limits on comments so they must be a minimum of 32 characters, and a maximum of 1024 characters. I've also added a limit to the screenshot links as well, they can be up to a maximum of 255 characters.

See commit `f5bb6c7` for the details.

File: index.php
Comment (WittyWhiscash) [reply index.php]: Good point. I've adjusted the increment to happen at the database level.

See commit `bd56692` for the details.

File: admin.php
Comment (WittyWhiscash) [reply admin.php]: That's a good idea. I've added a field on the comment in the database which gives administrators the official override on a comment so no matter how many times people will spam the report button, if you override, that's the final say.

See commit `b01dfb2` for the details.

**Review by 14ercooper — APPROVED** [review]

**Comment by 14ercooper** [comment]
Merged. Some potential future improvements if you (or anybody else) want something to do:
- Allow admins to list all comments and not just flagged comments for deletion/etc
- Add "rating" to the sort order for Browse, and somehow to Advanced Search
- Add a new "Top Rated" maps page, similar to Popular/Recently Added
- Don't show the "report comment" button on admin-approved comments

**Pull request merged**
